### ISSUE TITLE
Non-deterministic Slice= assignment

### ISSUE THREAD

**Issue by Werkov** [systemd/systemd#20182]
**systemd version the issue has been seen with**
 > 249

**Used distribution**
 > openSUSE Tumbleweed (mkosi image)

**Linux kernel version used** (`uname -a`)
 > `5.13.0-rc7`

**Expected behaviour you didn't see**
 > The slice unit configured for a unit follows deterministic rules (e.g. last assignment wins as with <v249).

**Unexpected behaviour you saw**
 > Slice assignments are evaluate in random order

```
image:~ # cat /etc/systemd/system/foo.service
[Service]
Slice=bar.slice
Slice=bar2.slice
ExecStart=/usr/bin/sleep inf

image:~ # cat /etc/systemd/system/goo@.service
[Service]
Slice=car.slice
Slice=car2.slice
ExecStart=/usr/bin/sleep inf

Control group /:
-.slice
├─bar.slice 
│ └─foo.service 
│   └─174 /usr/bin/sleep inf
├─car2.slice 
│ └─goo@alpha.service 
│   └─198 /usr/bin/sleep inf
[...]
```

**Steps to reproduce the problem**
> - create a unit file (foo.service) with multiple `Slice=` assignments,
> - `systemctl daemon-reload`
> - `systemctl start foo.service`
> - watch the service lands where in the tree 

I noticed this when looking at d219a2b07cc5dc8ffd5010f08561fab2780d8616. At first, I suspected that the custom slice assignment may not be honored since it would share drawing drum with the default assignment. My quick test didn't show that but I also can't see why it couldn't happen with other unit/slice names (IIUC, hashmap ordering is pseudorandom). The confirmed issue is that it changes semantics of multiple `Slice=` assignments (the last one used to overwrite previous), custom slices _likely_ work. Nevertheless, the new deps implementation should allow replacing for one-of-a-kind relations (such as slice).

**Comment by poettering** [comment]
Closely related to #20272. (The latter is about units created via bus calls, and this one is about those defined via unit files)

**Comment by intelfx** [comment]
This (899acf5) is supposed to also fix #20272, right?

**Comment by Werkov** [comment]
Yes, it boils down to `unit_set_slice()` with the last-one-wins semantics.
Thanks for negative reproducing, I think you can close #20272.

**Comment by poettering** [comment]
Closed #20272 now. @intelfx thanks for reporting back
